src/tools/imgtool/main.cpp

```diff
@@ -742,77 +739,64 @@ static int cmd_listfilters(const struct command *c, int argc, char *argv[])
 	return 0;
 }
 
-static void listoptions(const option_guide *opt_guide, const char *opt_spec)
+static void listoptions(const util::option_guide &opt_guide, const char *opt_spec)
 {
-	char opt_name[32];
-	const char *opt_desc;
-	util::option_resolution::range range[32];
-	char range_buffer[512];
-	char buf[32];
-	int i;
-
-	assert(opt_guide);
+	util::option_resolution resolution(opt_guide);
+	resolution.set_specification(opt_spec);
 
 	fprintf(stdout, "Option           Allowed values                 Description\n");
 	fprintf(stdout, "---------------- ------------------------------ -----------\n");
 
-	while(opt_guide->option_type != OPTIONTYPE_END)
+
+	for (auto iter = resolution.entries_begin(); iter != resolution.entries_end(); iter++)
 	{
-		range_buffer[0] = 0;
-		snprintf(opt_name, ARRAY_LENGTH(opt_name), "--%s", opt_guide->identifier);
-		opt_desc = opt_guide->display_name;
+		const auto &entry = *iter;
+		std::stringstream description_buffer;
 
-		/* is this option relevant? */
-		if (!strchr(opt_spec, opt_guide->parameter))
-		{
-			opt_guide++;
-			continue;
-		}
+		std::string opt_name = string_format("--%s", entry.identifier());
+		const char *opt_desc = entry.display_name();
 
-		switch(opt_guide->option_type) {
-		case OPTIONTYPE_INT:
-			util::option_resolution::list_ranges(opt_spec, opt_guide->parameter,
-				range, ARRAY_LENGTH(range));
+		// is this option relevant?
+		if (!strchr(opt_spec, entry.parameter()))
+			continue;
 
-			for (i = 0; range[i].max >= 0; i++)
+		switch (entry.option_type())
+		{
+		case util::option_guide::entry::option_type::INT:
+			for (const auto &range : entry.ranges())
 			{
-				if (range[i].min == range[i].max)
-					snprintf(buf, ARRAY_LENGTH(buf), "%i", range[i].min);
-				else
-					snprintf(buf, ARRAY_LENGTH(buf), "%i-%i", range[i].min, range[i].max);
+				if (!description_buffer.str().empty())
+					description_buffer << "/";
 
-				if (i > 0)
-					strncatz(range_buffer, "/", sizeof(range_buffer));
-				strncatz(range_buffer, buf, sizeof(range_buffer));
+				if (range.min == range.max)
+					util::stream_format(description_buffer, "%d", range.min);
+				else
+					util::stream_format(description_buffer, "%d-%d", range.min, range.max);
 			}
 			break;
 
-		case OPTIONTYPE_ENUM_BEGIN:
-			i = 0;
-			while(opt_guide[1].option_type == OPTIONTYPE_ENUM_VALUE)
+		case util::option_guide::entry::option_type::ENUM_BEGIN:
+			for (auto enum_value = entry.enum_value_begin(); enum_value != entry.enum_value_end(); enum_value++)
 			{
-				if (i > 0)
-					strncatz(range_buffer, "/", sizeof(range_buffer));
-				strncatz(range_buffer, opt_guide[1].identifier, sizeof(range_buffer));
-
-				opt_guide++;
-				i++;
+				if (!description_buffer.str().empty())
+					description_buffer << "/";
```

Comment:
Using the C string overload is far less efficient than the character overload when you have a single character.